### PR TITLE
[#3848] Widen CSP form-action with active SSO IdP origins

### LINKED ISSUE #3848: SSO: CSP `form-action 'self'` blocks IdP redirect on Chromium (v0.26 regression via otto 2.5+)
## Summary

Since v0.26.0-rc1, SSO login is blocked in Chromium-family browsers (Chrome, Edge, Vivaldi, Brave) at the very first step: the browser refuses to follow the redirect from `POST /auth/sso/:provider` to the identity provider, because the emitted Content-Security-Policy contains `form-action 'self'`. This is the **primary cause of #3836** (Bug A). The account-linking refusal diagnosed in the same thread (Bug B) is tracked separately in #3838.

The failure is UI-silent: clicking the SSO button does nothing. Server logs show `POST /auth/sso/entra → 302` and then silence — the browser blocks the redirect client-side, so the callback never happens and nothing appears in IdP sign-in logs.

## Root cause

Three facts stack:

1. **SSO sign-in is a real DOM form POST** (`src/shared/utils/sso.ts`), deliberately not a fetch, because the endpoint 302s the browser to the IdP and an XHR can't follow that cross-origin redirect. The form action is relative, so the POST itself is same-origin and legal under `form-action 'self'`.
2. **Chromium enforces `form-action` against the redirect chain**, not just the form action URL (long-standing spec dispute, w3c/webappsec-csp#8). The 302 to `login.microsoftonline.com` (or any IdP origin) trips the directive. Firefox does not enforce on redirects, which is why the bug is browser-dependent.
3. **otto 2.5.0 started actually emitting the CSP header.** Before otto 2.5, our CSP config was enabled but the header never went out (the #3558 gap). otto's generated policy hardcodes `form-action 'self'` with no SSO awareness.

### Version matrix (otto per Gemfile.lock at each tag)

| OTS tag | otto | SSO in Chromium |
|---|---|---|
| v0.25.11 | 2.2.0 | works (no CSP header emitted) |
| v0.26.0-rc0 | 2.3.1 | works (no CSP header emitted) |
| v0.26.0-rc1 | 2.5.0 | **broken** (header emitted, `form-action 'self'`) |
| v0.26.0-rc2 / v0.26.0 / v0.26.1 | 2.6.0 | **broken** |

CSP defaults ON (`etc/defaults/config.defaults.yaml` — `CSP_ENABLED != 'false'`); emission happens in `apps/web/core/middleware/request_setup.rb#emit_csp_header` via `Otto::Security::CSP::Writer.apply`.

## Evidence

- Reproduced locally on two providers and two Chromium browsers: Chrome 150 (`oidc` provider) and Vivaldi (`entra` provider). Console: `Sending form data to 'https://<host>/auth/sso/entra' violates the following Content Security Policy directive: "form-action 'self'". The request has been blocked.` The POST shows `net::ERR_ABORTED`; the page stays on `/signin`.
- Note the console message names the **same-origin form action URL, not the IdP** — CSP violation reporting intentionally reports the pre-redirect URL to avoid leaking redirect targets. Don't let that mislead triage.
- Firefox control: the identical flow completes end-to-end (no redirect-chain enforcement).
- Matches both field reports in #3836: RedEchidnaUK (Entra, address bar never leaves the OTS domain, nothing in MS sign-in logs) and syphernl (generic OIDC, same console violation).

## Fix plan (app-side only — no otto release required)

otto 2.6.0 already has the seam: `Otto::Security::Config#csp_directive_overrides` / `#merge_csp_directives`, and `enable_csp_with_nonce!(debug:, directives:)` accepts overrides directly. We call it bare today in `apps/web/core/application.rb#build_router`:

```ruby
router.enable_csp_with_nonce!(debug: OT.debug?)
```

Plan:

1. **Origin derivation helper** (new, shared): compute the set of SSO IdP origins from the same ENV the auth app's provider registration reads (`apps/web/auth/config/features/omniauth.rb`), so web-core's router build doesn't depend on auth-app boot order:
   - Entra (`ENTRA_TENANT_ID` et al configured): `https://login.microsoftonline.com`; support sovereign clouds if a non-default authority is configured.
   - Google: `https://accounts.google.com`
   - GitHub: `https://github.com`
   - Generic OIDC (`OIDC_ISSUER`): origin of the issuer URL as best effort. Caveat: the discovered `authorization_endpoint` origin can differ from the issuer origin, and discovery happens at request time, not boot.
   - **Env override / catch-all** (e.g. `SSO_FORM_ACTION_ORIGINS`, space-separated origins): covers the OIDC issuer≠authorization-endpoint case, federated Entra hops to customer ADFS mid-chain, and per-tenant org SSO where issuers aren't knowable at boot (placeholder-provider registration path).
2. **Wire it in**: `router.enable_csp_with_nonce!(debug: OT.debug?, directives: { 'form-action' => "'self' #{origins.join(' ')}" })` — only when at least one origin is derived; with no SSO configured the policy stays exactly as today.
3. Do **not** mutate the shared security config per request; boot-time global origins only in this pass. Per-tenant per-request tightening is a possible follow-up, not v0.26.2.

## Acceptance criteria

- [ ] With Entra configured, emitted CSP contains `form-action 'self' https://login.microsoftonline.com` (analogously per provider); with no SSO configured, policy is unchanged.
- [ ] `SSO_FORM_ACTION_ORIGINS` env var merges additional origins; documented.
- [ ] Chromium-based regression coverage: SSO button click navigates off-origin (or produces no CSP form-action violation). Must run Chromium — Firefox cannot regress on this.
- [ ] Unit spec for the origin-derivation helper (per provider, override var, invalid URL tolerance).
- [ ] Request/middleware spec asserting the emitted header.
- [ ] Docs: `docs/authentication/omniauth-sso.md` section on CSP + SSO origins; CHANGELOG upgrade note naming the v0.26.0/.1 breakage and the `CSP_ENABLED=false` interim workaround.

## Workaround (until fixed)

`CSP_ENABLED=false` — drops the entire CSP header. Restore it after upgrading to the fix release.

## Related

- #3836 — reporter thread (this is Bug A / primary cause)
- #3838 — Bug B (H-3 account-linking refusal), independent work
- #3558 — why the CSP header only began emitting at otto 2.5
- Deployment: header-only behavior change; rollback = revert the commit. `CSP_ENABLED` remains the operator kill switch.

**Complexity**: small — one helper + one call-site kwarg + tests/docs. The care is in origin derivation edge cases, not the wiring.

### PR DESCRIPTION
**Pull request by delano**
## Summary

Fixes #3848 — SSO login broke on Chromium-family browsers (Chrome, Edge, …) in **v0.26.0-rc1**.

otto 2.5 emits a CSP containing `form-action 'self'`. Chromium enforces `form-action` **across the entire redirect chain**, so the SSO form-POST to `/auth/sso/:provider` that 302-redirects to the IdP was blocked — clicking a provider button did nothing. Firefox only checks the initial (same-origin) form target, so it was unaffected, which is why the bug reproduced in one browser family and not the other.

## Fix

At boot, add each **active** SSO provider's IdP origin to the `form-action` directive:

```ruby
sso_origins = Onetime.auth_config.sso_form_action_origins
unless sso_origins.empty?
  router.security_config.merge_csp_directives('form-action' => "'self' #{sso_origins.join(' ')}")
end
```

- **Gating reuse:** `sso_form_action_origins` derives origins from the *same* `provider_definitions` gate as `#sso_providers` (SSO enabled **and** required env vars present), so `form-action` can never drift from the providers that actually register.
- **Static** origins for Entra (`login.microsoftonline.com`), Google (`accounts.google.com`), GitHub (`github.com`); **derived** origin from `OIDC_ISSUER` for generic OIDC.
- **Fail-safe:** a malformed/schemeless/hostless issuer is skipped, never raised (`filter_map` + `URI` guard).
- **No-op** when no SSO provider is configured (policy stays byte-identical) and **inert** when CSP is disabled (`site.security.csp.enabled`).

### Operator override

`SSO_FORM_ACTION_ORIGINS` (space-separated) merges in unconditionally for cases the auto-derivation can't cover:

- Sovereign / national clouds (e.g. `login.microsoftonline.us`)
- An OIDC issuer whose `authorization_endpoint` origin differs from `OIDC_ISSUER`
- Org-level SSO whose issuers are unknown at boot

**Interim workaround** for un-upgraded installs: `CSP_ENABLED=false` drops the CSP header entirely.

## Tests

- **Unit** — `spec/unit/onetime/config/auth_config_sso_form_action_origins_spec.rb`: provider gating, OIDC derivation, port handling, malformed-issuer tolerance, override merge/dedup, feature-disabled short-circuit.
- **Middleware** — `apps/web/core/spec/middleware/csp_form_action_sso_spec.rb`: emitted `form-action` directive with/without injected origins.
- **E2E** — `e2e/auth/sso-csp-form-action.spec.ts`.

20/20 Ruby examples green locally.

## Also updated

- `.env.reference` — documents `SSO_FORM_ACTION_ORIGINS`
- `docs/authentication/per-install-sso.md` — troubleshooting section + env table row
- `changelog.d/…` — scriv fragment

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: spec/support/auth_mode_helpers.rb [hunk 1]
```diff
@@ -220,6 +220,14 @@ def session
...
       self
     end
+
+    # Mirrors Onetime::AuthConfig#sso_form_action_origins. Tests don't set
+    # provider env vars, so this only reflects the SSO_FORM_ACTION_ORIGINS
+    # override — enough for router/CSP boot specs that just need the method
+    # to respond without raising.
+    def sso_form_action_origins
+      ENV.fetch('SSO_FORM_ACTION_ORIGINS', '').to_s.split
+    end
```
[the remark this answers is not in the document]
Comment (delano) [reply spec/support/auth_mode_helpers.rb:234]: Done in 816fa36e98 — the stub now applies `.uniq`, matching `AuthConfig#sso_form_action_origins`. Token http(s)/injection validation is intentionally left to `AuthConfig#origin_from_url` (single source of truth) rather than duplicated in the stub; boot specs don't set malformed overrides.

**Comment by Pant1f3r** [comment]
Nice

**Comment by Pant1f3r** [comment]
Ok
